packages/test-utils/src/run.ts

```diff
@@ -72,24 +77,31 @@ function gatherEntities(result: any): Entity[] {
   }
 }
 
-async function mapResultToOriginalEm<R>(em: EntityManager, result: R): Promise<R> {
+async function mapResultToOriginalEm<R>(em: EntityManager, result: R, opts?: RunOptions): Promise<R> {
   const newEmEntities = gatherEntities(result);
   // load any entities that don't exist in the original em
   await Promise.all(
     newEmEntities
-      .filter(
-        (e) =>
-          !em.findExistingInstance(
-            e.idTaggedMaybe ?? fail("Joist 'run' returned an unsaved entity; are you missing an em.flush?"),
-          ),
-      )
+      .filter((e) => {
+        if (e.idTaggedMaybe === undefined) {
+          if (opts?.expectUnsavedEntities === true) return false;
+          fail("Joist 'run' returned an unsaved entity; are you missing an em.flush?");
+        }
+        return !em.findExistingInstance(e.idTaggedMaybe);
+      })
```

Comment:
issue: this code would actually fail in `load` if `assignNewIds` had been called during the test run, but no flush had been called.  The previous implementation had the same fault, but since it never intended to support this behavior it wasn't an issue.  You should use `isNewEntity` instead.

nit: this implementation is getting a little convoluted, it seems more readable to reformat it.  Incorporating @stephenh's naming suggestion:



```typescript
const { dryRun = false } = opts ?? {};
if (!dryRun && newEmEntities.some((e) => e.isNewEntity) )
  fail("Joist 'run' returned an unsaved entity; are you missing an em.flush?");

await Promise.all(
    newEmEntities
      .filter((e) => !e.isNewEntity)
      .filter((e) => !em.findExistingInstance(e.idTagged))
      .map((e) => em.load(e.idTagged))
);
```